bzzeth/bzzeth_test.go

```diff
@@ -223,3 +260,170 @@ func isPeerDisconnected(id enode.ID, b *BzzEth) (p *Peer) {
 	}
 	return
 }
+
+func newBlockHeaderExchange(tester *p2ptest.ProtocolTester, peerID enode.ID, requestID uint32, offered *NewBlockHeaders, wanted [][]byte) error {
+	return tester.TestExchanges(
+		p2ptest.Exchange{
+			Label: "NewBlockHeaders",
+			Triggers: []p2ptest.Trigger{
+				{
+					Code: 1,
+					Msg:  offered,
+					Peer: peerID,
+				},
+			},
+			Expects: []p2ptest.Expect{
+				{
+					Code: 2,
+					Msg: GetBlockHeaders{
+						ID:     requestID,
+						Hashes: wanted,
+					},
+					Peer: peerID,
+				},
+			},
+		})
+}
+
+func blockHeaderExchange(tester *p2ptest.ProtocolTester, peerID enode.ID, requestID uint32, wantedData []rlp.RawValue) error {
+	return tester.TestExchanges(
+		p2ptest.Exchange{
+			Label: "BlockHeaders",
+			Triggers: []p2ptest.Trigger{
+				{
+					Code: 3,
+					Msg: BlockHeaders{
+						ID:      requestID,
+						Headers: wantedData,
+					},
+					Peer: peerID,
+				},
+			},
+		})
+}
+
+// Test bzzeth full eth node sends new block header hashes
+// respond with a GetBlockHeaders requesting headers falling into the proximity of this node
+// Also test two other conditions
+// - If a header is already present in localstore, dont request it in GetBlockHeaders
+// - If a unsolicited header is received, dont store it on localstore
+// Apart from that it also tests if all the headers are delivered and stored in localstore
+func TestNewBlockHeaders(t *testing.T) {
+	prvKey, netstore, cleanup := newTestNetworkStore(t)
+	defer cleanup()
+
+	// bzz pivot - full eth node peer
+	// NewBlockHeaders trigger, expect
+	tester, _, teardown, err := newBzzEthTester(t, prvKey, netstore)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer teardown()
+
+	offered := make(NewBlockHeaders, 256)
+	for i := 0; i < len(offered); i++ {
+		offered[i].Hash = common.BytesToHash(crypto.Keccak256([]byte{uint8(i)}))
+		offered[i].Number = uint64(i)
+	}
+
+	// redefine wantHeadeFunc for this test
+	wantedIndexes := []int{1, 2, 3, 5, 8, 13, 21, 34, 55, 89, 144, 233}
+	IgnoreIndexes := []int{77}
+	wantHeaderFunc = func(hash []byte, _ *network.Kademlia) bool {
+		for _, i := range wantedIndexes {
+			if bytes.Equal(hash, offered[i].Hash.Bytes()) {
+				return true
+			}
+			// Add the ignore headers (headers in localstore already) to the valid list
+			if bytes.Equal(hash, offered[IgnoreIndexes[0]].Hash.Bytes()) {
+				return true
+			}
+		}
+		return false
+	}
+
+	wanted := make([][]byte, len(wantedIndexes))
+	wantedData := make([]rlp.RawValue, len(wantedIndexes)+1)
+	for i, w := range wantedIndexes {
+		wanted[i] = crypto.Keccak256([]byte{uint8(w)})
+		wantedData[i] = rlp.RawValue{uint8(w)}
+	}
+
+	// overwrite newRequestIDFunc to be deterministic
+	defer func(f func() uint32) {
+		newRequestIDFunc = f
+	}(newRequestIDFunc)
+
+	newRequestIDFunc = func() uint32 {
+		return 42
```

Comment:
this is very difficult to read and review. there's a whole bunch of indexes here which affect the test logic and they are scattered and potentially changed, please either change this to a table driven test where the inputs and outputs of the test are cleared in advance or lay out the entire test case beforehand, including all sorts of arbitrary integer values.
thanks for all the fish :fish:

Reply:
@acud  Well... you could have avoided the "Thanks to all the fish" thing.. its not funny.

Reply:
first.. there is only two indexes ... wantedIndex and IgnoreIndex.
The ways it is  done like this is to pass the arguments separately to the protocol check function.I agree that this is not very clear in the beginning... if you spend some time.. you will get it..